Remove Font Renderer + CP Replace + libagg

adamharrison opened the pull request:
Removed libagg, and the C++ font renderer, as per the discussion in #409.

Also removed cp_replacements.c, and wrote a plugin to replace the only thing in the core that actually used it for anything (which shouldn't be particularly intensive).

I'm still tweaking the meson build config, but can be built with the following line: 

```
gcc -O3 `find src/ -iname "*.c"` -o lite-xl `pkg-config --cflags sdl2 freetype2 lua5.2` -Isrc `pkg-config --libs reproc sdl2 freetype2 lua5.2` -lm -lpthread -lpcre2-8
```

adamharrison commented:
OK, I think I got meson working.

franko commented:
Thank you @adamharrison for doing this work, I really appreciate.

That's a massive simplification for the project and something I wanted to do since a long time. You did an amazing work!

Let's give some stats:

- binary size -123 kb (-5 %)
- compile time (optimized build on windows) -52 %, from 16.09s to 7.621s, probably bigger improvement on linux due to faster filesystem operations
- project complexity, -50% (I used my nose for this estimation)
- third party library dependencies, -17% (-libagg)
- number of line of codes (only compiled source code) -51% (-2577 lines)
- do things like everyone else, +infinity% (we did nothing like everyone else)

franko commented on src/api/regex.c:
```diff
@@ -101,17 +101,11 @@ int luaopen_regex(lua_State *L) {
-  lua_setfield(L, -2, "NOTEMPTY");
-  lua_pushnumber(L, PCRE2_NOTEMPTY_ATSTART);
-  lua_setfield(L, -2, "NOTEMPTY_ATSTART");
+  API_CONSTANT_DEFINE(L, -1, "ANCHORED", PCRE2_ANCHORED);
```
These changes here seems completely unrelated to the topic of this PR.

adamharrison replied:
Ah yes, sorry; this bled in from my other changes to process; I'll revert these.

franko commented:
You removed the cp_replacements code but I implemented this code because otherwise the "drawwhitespace" plugin was increasing a lot the CPU usage. I didn't check again your new implementation in Lua, maybe it is better but I don't know at the moment.

I thought this is a plugin a lot of people were using so it was important to make it perform well.

franko commented:
You removed also the FontDesc structure I introduced so probably retina display is broken on macOS but I need to check.

I did this change to support retina displays on macOS. What happen in that on macOS you request a font at size 13 but in reality you may need to render it at size 13 or at size 26 or at size 52 because of the scaling factor. The scaling is not known when you are londing the font but you know it only when you create the actual window.

franko commented:
You removed subpixel positioning right ? This may degrade rendering quality slightly in term of horizontal spacing but may still be ok.

If I understand correctly the blending operation performed on glyphs is the same.


New code (for subpixel):
```c
            SDL_Color src = { *source_pixel++, *source_pixel++, *source_pixel++ };
            r = color.r * src.r + dst.r * (255 - src.r) + 127;
            r = DIVIDE_BY_255(r);
            g = color.g * src.g + dst.g * (255 - src.g) + 127;
            g = DIVIDE_BY_255(g);
            b = color.b * src.b + dst.b * (255 - src.b) + 127;
            b = DIVIDE_BY_255(b);
```

Old code:
```c++
    const int pixel_index[3] = {Order::R, Order::G, Order::B};
    for (unsigned cx = 0; cx < len; cx += 3)
    {
        for (int i = 0; i < 3; i++) {
            const unsigned cover_value = covers[cx + i];
            const unsigned alpha = (cover_value + 1) * (c.a + 1);
            const unsigned src_col = *(p + pixel_index[i]);
            *(p + pixel_index[i]) = (((rgb[i] - src_col) * alpha) + (src_col << 16)) >> 16;
        }
        // Leave p[3], the alpha channel value unmodified.
        p += 4;
    }
```

I am not sure, I need to write down the math in a paper to check if they are doing the same thing or not.

franko commented:
You may consider adding these modifications in your branch. It correctly disable the "whitespace" plugin (by default) and adapt font size to be similar to what we have now in the master branch.

```diff
 config.transitions = true
 config.animation_rate = 1.0
 config.blink_period = 0.8
-config.draw_whitespace = false
 config.borderless = false
 config.tab_close_button = true

@@ -33,5 +32,6 @@ config.plugins = {}

 config.plugins.trimwhitespace = false
 config.plugins.lineguide = false
+config.plugins.drawwhitespace = false

 return config
diff --git a/data/core/style.lua b/data/core/style.lua
index 7cf16eb..26187e9 100644
--- a/data/core/style.lua
+++ b/data/core/style.lua
@@ -21,11 +21,11 @@ style.tab_width = common.round(170 * SCALE)
 --
 -- On High DPI monitor or non RGB monitor you may consider using antialiasing grayscale instead.
 -- The antialiasing grayscale with full hinting is interesting for crisp font rendering.
-style.font = renderer.font.load(DATADIR .. "/fonts/FiraSans-Regular.ttf", 13 * SCALE)
-style.big_font = style.font:copy(40 * SCALE)
-style.icon_font = renderer.font.load(DATADIR .. "/fonts/icons.ttf", 14 * SCALE, {antialiasing="grayscale", hinting="full"})
-style.icon_big_font = style.icon_font:copy(20 * SCALE)
-style.code_font = renderer.font.load(DATADIR .. "/fonts/JetBrainsMono-Regular.ttf", 13 * SCALE)
+style.font = renderer.font.load(DATADIR .. "/fonts/FiraSans-Regular.ttf", 15 * SCALE)
+style.big_font = style.font:copy(46 * SCALE)
+style.icon_font = renderer.font.load(DATADIR .. "/fonts/icons.ttf", 16 * SCALE, {antialiasing="grayscale", hinting="full"})
+style.icon_big_font = style.icon_font:copy(23 * SCALE)
+style.code_font = renderer.font.load(DATADIR .. "/fonts/JetBrainsMono-Regular.ttf", 15 * SCALE)

 style.background = { common.color "#2e2e32" }
 style.background2 = { common.color "#252529" }
```

to ease testing.

adamharrison commented:
> You removed the cp_replacements code but I implemented this code because otherwise the "drawwhitespace" plugin was increasing a lot the CPU usage. I didn't check again your new implementation in Lua, maybe it is better but I don't know at the moment.

> I thought this is a plugin a lot of people were using so it was important to make it perform well.

I concur that it should perform well; but there was a not-insignificant amount of rather obtuse C code behind this for a single feature. I did some simple profiling here, and it doesn't look like it affects CPU% use on my machine basically in any detectable way. After all, the tokenizer is also making one rendering call per colored bit of a line, and the rendering calls are not usually the expensive ones, thanks to the rencache.

That said, if there is a worry about performance, I have some ideas on how to improve `DocView` in general, rather than the cp_replace stuff specifically.

> You removed also the FontDesc structure I introduced so probably retina display is broken on macOS but I need to check.

> I did this change to support retina displays on macOS. What happen in that on macOS you request a font at size 13 but in reality you may need to render it at size 13 or at size 26 or at size 52 because of the scaling factor. The scaling is not known when you are londing the font but you know it only when you create the actual window.

So this is confusing to me; it was my understanding that this was what `SCALE` was primarily for; it's computed from the DPI of the display; and we scale the fonts with this number. So if you have a retina display on mac with a high dpi, we already know this from `get_scale`, no?

Although, I do now see this in the function:

```
static double get_scale(void) {
#ifdef __APPLE__
    return 1.0;
#else
  float dpi = 96.0;
  SDL_GetDisplayDPI(0, NULL, &dpi, NULL);
  return dpi / 96.0;
#endif
}
```

In this case, I assume SDL_GetDisplayDPI is incorrect on Macs. I'll investigate this.

> You removed subpixel positioning right ? This may degrade rendering quality slightly in term of horizontal spacing but may still be ok.

No; I didn't. Subpixel position is still present; all `draw_text` calls take a `float`, for the x-coordinate which represents the subpixel position as the fractional part of the `float`. Drawing something at 10.5 will use the closest available available subpixel shifted bitmap as per this line:

`int bitmap_index = font->subpixel ? (int)(fmod(pen_x, 1.0) * SUBPIXEL_BITMAPS_CACHED) : 0;`

> You may consider adding these modifications in your branch. It correctly disable the "whitespace" plugin (by default) and adapt font size to be similar to what we have now in the master branch.

Good point, I'll make sure to throw that in as well.

adamharrison commented:
> In this case, I assume SDL_GetDisplayDPI is incorrect on Macs. I'll investigate this.

Oh, I see. You have it here:

```
#ifdef LITE_USE_SDL_RENDERER
static int query_surface_scale(RenWindow *ren) {
  int w_pixels, h_pixels;
  int w_points, h_points;
  SDL_GL_GetDrawableSize(ren->window, &w_pixels, &h_pixels);
  SDL_GetWindowSize(ren->window, &w_points, &h_points);
  /* We consider that the ratio pixel/point will always be an integer and
     it is the same along the x and the y axis. */
  assert(w_pixels % w_points == 0 && h_pixels % h_points == 0 && w_pixels / w_points == h_pixels / h_points);
  return w_pixels / w_points;
}

static void setup_renderer(RenWindow *ren, int w, int h) {
  /* Note that w and h here should always be in pixels and obtained from
     a call to SDL_GL_GetDrawableSize(). */
  if (ren->renderer) {
    SDL_DestroyTexture(ren->texture);
    SDL_DestroyRenderer(ren->renderer);
  }
  ren->renderer = SDL_CreateRenderer(ren->window, -1, 0);
  ren->texture = SDL_CreateTexture(ren->renderer, SDL_PIXELFORMAT_BGRA32, SDL_TEXTUREACCESS_STREAMING, w, h);
  ren->surface_scale = query_surface_scale(ren);
}
#endif```

This code is all called *before* setup. Could we not simply bundle this into `get_scale`?

franko commented:
In this PR you are doing several changes at once:

1. replace libagg + freetype C++ code with a simple C freetype only approach to render glyphs
2. remove the cp_replacements C code and adopt a Lua only approach for the scale plugin
3. remove FontDesc to go back to RenFont removing unknowingly support for retina display on macOS
4. replace the system with integer subpixel coordinates with a system with floating point pixel coordinates (on macOS actually there are points coordinates and subpixel coordinates).

Cumulating several changes like this in a single PR make it more difficult to integrate it without introducing regressions.

Ideally I would like to have a PR only with (1). It would be less risky and easy to accept for me.

Otherwise, I am ready to give up about (2).

For the point (4) it may be ok but don't know how it interacts with macOS with retina display but your proposition may be okay. It was just better to evaluate it separately.

As for (3) you are thinking now to find a solution to the problem I already solved introducing FontDesc. It seems to me you removed this feature "by accident" because you didn't know it was used for macOS to support retina displays.

Guldoman commented on src/renderer.c:
```diff
@@ -59,213 +65,206 @@ static const char* utf8_to_codepoint(const char *p, unsigned *dst) {
+          break;
+        unsigned int* destination_pixel = (unsigned int*)&destination_pixels[surface->pitch * targetY + start_x * bytes_per_pixel];
+        unsigned char* source_pixel = &source_pixels[line * set->surface[bitmap_index]->pitch + metric->x0 * (font->subpixel ? 3 : 1)];
+        for (int x = glyphStart; x < glyph_end; ++x) {
```
I think you should check if `x` is beyond the clip here too.

adamharrison replied:
This is a good point. We shouldn't draw the glyph past the clip. Will modify.

adamharrison replied:
Finally fixed this.

adamharrison commented:
>  In this PR you are doing several changes at once:

> replace libagg + freetype C++ code with a simple C freetype only approach to render glyphs
> remove the cp_replacements C code and adopt a Lua only approach for the scale plugin
> remove FontDesc to go back to RenFont removing unknowingly support for retina display on macOS
> replace the system with integer subpixel coordinates with a system with floating point pixel coordinates (on macOS > actually there are points coordinates and subpixel coordinates).
> Cumulating several changes like this in a single PR make it more difficult to integrate it without introducing regressions.

> Ideally I would like to have a PR only with (1). It would be less risky and easy to accept for me.

Agreed. That's the minimal PR.

> Otherwise, I am ready to give up about (2).

I think this is a good thing to include; otherwise I'd have to write new code to include this, rather than include old code. Definitely doable, but if I were to do this, I'd like to see counter-evidence against my current profiling that it is really CPU intensive that would impact people using `lite-xl`.

> For the point (4) it may be ok but don't know how it interacts with macOS with retina display but your proposition may be okay. It was just better to evaluate it separately.

I don't know either about this one. I'd have to work on a retina display to determine (which I do have access to). I'll have to get back to you on this one. I don't see logically why it'd be different though.

> As for (3) you are thinking now to find a solution to the problem I already solved introducing FontDesc. It seems to me you removed this feature "by accident" because you didn't know it was used for macOS to support retina displays.

This however, I do think we have an obvious solution. It does depend on how those coordinates change on MacOS Retina. As I understand it, this is a single integer determined a bootup about MacOS, right? So I have a few questions:

1. Does this integer change on resizing the window?
2. If not, does the integer *ever* change on a particular system post-bootup?
3. It this integer apply to any system other than MacOS?

If this answer to all these three questions is no, I think we can safely bundle it into `SCALE`. If not, it might be more complex, and may make more sense to look at `FontDesc` though that is quite a lot of code, that I think can be simplified, even if 1)&2) are not correct. Either way, I don't think we need extra structure to deal with it, perhaps only an extract if statement or two, though I do want to understand this fully before I commit.

That said, I do have a friend's laptop to verify with retina display, so in the worst case, I can just fix it on his laptop and see; but ideally we can find something together.

adamharrison commented:
So, just a follow up here; should have access to a macbook with retina display this weekend; so should be able to address points 1&2 up there shortly.

As a bonus though, I just threw in bolding and italics, and it seems to be working quite well, so that's another plus for this.

adamharrison commented:
Also, got in underline.

adamharrison commented:
OK; finally managed to secure a retina display. Will look into it tomorrow.

adamharrison commented:
Well, uh, I got a Mac. But I need to update it, because Xcode won't install without that (which I apparently need for meson/build tools). And it says that that'll take ~4 hours. And the battery's draining. I plug it in, and it tells me it's plugged in, but it also says it's not charging, and the battery's still draining. So I'm not sure if I'll actually be testing this, lol. I guess we'll see.

adamharrison commented:
OK! So, despite the fact that I can't charge the Mac, I *think* I got everything working, just as the battery hit 25%. I can potentially get another one, but everything looked good when I made those changes on the local copy.

So, I'm pretty sure I tested point 1) above; and it doesn't change. Point 2), however, I'm unsure. If it is the case that it changes; it's a pretty straightforward recomputation; we simply store the scale at which we generated the font, and run something along the lines of:

```
FT_Set_Pixel_Sizes(font->face, 0, (int)(size*surface_scale));
for (int i = 0; i < MAX_GLYPHSET; ++i) {
  if (font->sets[i])
    font_free_glyphset(font->sets[i]);
}
font->scale = surface_scale;
```

And that'd solve the problem. But I'm not sure it's necessary.

adamharrison commented:
I would love it, if anyone who has a mac could just take a quick peek around this branch to verify there aren't any major issues. Not sure if I can get my hands on another one easily at this point.

adamharrison commented:
Managed to get my hands on a mac, revealed a few more problems which have now been fixed. I played with it for a good 20 minutes, and everything seems fine on both Linux and Mac.

adamharrison commented:
OK, done. @franko if you can review at some point, I think this is ready to merge.

franko commented:
I am ready to merge this, I am just a little bit worried about macOS that we don't introduce regressions but I will test.

One of the remaining concern I have right now is that the font are too jarring to my eyes due to color fringes in the subpixel rendering. Here a comparison image, first two lines from standard lite-xl then above with the code from this branch. Below the same image scaled by 8 to see the pixels.

![image](https://user-images.githubusercontent.com/433545/134816837-ab64a584-0619-4bbc-bd2c-85922bcabb80.png)

![image](https://user-images.githubusercontent.com/433545/134816932-141b365c-b5be-4f8c-b31f-ecb0a06a5e82.png)

For this it could be my impression, I would like to have some feedback from other.

It would be more easy to accept this is the subpixel filtering was a little more conservative like the one we currently have. Maybe this can be tuned with some freetype library parameter.

franko commented:
I tested on macOS and I confirm everything seems fine with a standard display and with retina.

adamharrison commented:
I'll take a look and see if I can't tune down the color fringes, either in the library, or in the blend.

franko commented:
> I'll take a look and see if I can't tune down the color fringes, either in the library, or in the blend.

Thank you @adamharrison.

As far as I can tell the problem is not related to the blend operation but it happens before, when applying a subpixel filter so it happens inside the freetype library. If we fix this problem modifying the blending it would be an hack.

I theory we may ask the library to not do subpixel and just request grayscale alpha coverage with a x3 pixel resolution in the X axis and we may apply our own subpixel filter. For x3 resolution it can be done using a scaling affine matrix. For the subpixel filter in the original code it was in C++ but it can be easily adapted in C if we want to.

I advice first do a research if there are some freetype option we can use for the subpixel filter and use them if they are available. Using our own code would be the last resource.

franko commented:
If you want we may merge the branch as it is currently and work on this problem later.

adamharrison commented:
Agreed, it'd be a hack. But you can customize the subpixel filter; do you know what weights the original lite-xl renderer was using?

adamharrison commented:
What about this? Does that look a little better?

franko commented:
> Agreed, it'd be a hack. But you can customize the subpixel filter, so hacking the blend is likely not necessary. Do you know what weights the original lite-xl renderer was using?

Thank you, that's a great news, I will test the new settings.

Otherwise the values we are currently using are there:

```c++
class FR_Renderer {
public:
    // Conventional LUT values: (1./3., 2./9., 1./9.)
    // The values below are fine tuned as in the Elementary Plot library.

    FR_Renderer(bool hinting, bool kerning, bool subpixel, bool prescale_x) :
        m_renderer(hinting, kerning, subpixel, prescale_x),
        m_lcd_lut(0.448, 0.184, 0.092),
        m_subpixel(subpixel)
    { }
```

in `m_lcd_lut`, so 0.448, 0.184, 0.092 normalized to 1. If we call them a, b, c they are such a + 2*b + 2*c = 1. The coefficients are the weights of, respectively, the primary, secondary and tertiary channels.

Here the references I have used for the implementation in case you want to have a look:

https://www.grc.com/ctwhat.htm

and optionally for the AGG implementation:

https://franko.github.io/antigrain/research/font_rasterization/index.html#FONT_RASTERIZATION

By the way, these values have been optimized by me, I don't know if to someone else they look better than the freetype's default. For example I would like to have your personal impression about the look of the default freetype subpixel and the one we currently use in lite-xl.

franko commented:
I just discovered that the subpixel font positioning is all messed up with the new renderer.

I wanted to upload an image but I cannot at this moment for some reason. Please have a look at the treeview with FiraSans, the standard font. The spacing is all wrong for example in a file named "testing.txt" the "s" and the "t" will be very close to each other.

Now, I will just share with you what I think.

I think you were too much aggressive removing and redoing what I have done and adjusted over several months. I am thinking about the macOS retina display support, the subpixel positioning in rencache, and the codepoint replace stuff.

As you are a really smart programmer you are able to find good solutions, better than mine, but having to go through problems that have been already solved in the past is annoying and we risk adding regressions.

If you were more conservative doing just the replacement of libagg font rendering with native freetype calls that would be much more simple and less risky to introduce regressions. Other changes may have been done in separate PR. Cumulating all these changes all in one PR is making things complicated.

Now that you know what I think, it's okay. We can continue fixing all the issues that arise with this PR. I guess at some point everything will be okay.

franko commented:
In term of the subpixel rendering the new modification looks fine even if maybe not identical to the existing filter.

I only noticed the problem with character spacing. It seems to happen only on windows for some reason. I will double check and try to post the image.

adamharrison commented:
> I just discovered that the subpixel font positioning is all messed up with the new renderer.

> I wanted to upload an image but I cannot at this moment for some reason. Please have a look at the treeview with FiraSans, the standard font. The spacing is all wrong for example in a file named "testing.txt" the "s" and the "t" will be very close to each other.

Sure, I'll take a look at this; please upload an image when you can.

> I think you were too much aggressive removing and redoing what I have done and adjusted over several months. I am thinking about the macOS retina display support, the subpixel positioning in rencache, and the codepoint replace stuff.

> As you are a really smart programmer you are able to find good solutions, better than mine, but having to go through problems that have been already solved in the past is annoying and we risk adding regressions.

Give yourself some credit! I don't think I'm a better programmer than you; after all, I used your work as a base! And you've corrected me many times about the route I should be taking, and how subpixel positioning works; and for that, I'm very grateful. I think it would've taken a lot longer without you and your code here to guide.

I do agree; this is a relatively unnecessary PR in the general sense. It doesn't really add anything new (other than run-time generation of bold and italics and underlined fonts). However, as I understand it, one of the core philosophies of lite (and lite-xl), is that it is small in terms of its codebase and dependencies. Even if we're not adding much functionality with this, a PR like this helps lite-xl stick closer to that original goal (not to mention, we remove an entire language from the mix). So I consider it worthy, even if it doesn't add much, and has the risk of introducing regressions. We just have to QA things until we're sure that there are almost no regressions.

That said, I understand that this was your code that I'm removing, and I know that can sting. But I think in the general interests of the project; we shouldn't have a focus on code ownership. Like, for example, if someone were to find a way to do good regexes without pulling libpcre2; I'd welcome that change, even if they had to rip out my code; as generally it would be better for the project. Ditto with the subsyntax support. I think anything that *removes* lines of C code or dependencies is a good thing, regardless of whose lines they were.

> If you were more conservative doing just the replacement of libagg font rendering with native freetype calls that would be much more simple and less risky to introduce regressions. Other changes may have been done in separate PR. Cumulating all these changes all in one PR is making things complicated.

So, the issue is that if I were to leave much of this stuff in, I'd have to write extra code that we'd then end up deleting in later PRs. I agree with you that the smallest PR possible is generally better; but I generally don't want to write additional code that has to be tested, and fixed, if we're only going to be removing it later. I think it's tenable is absolutely critical situations, but given that this isn't a particularly critical PR, nor is it in any rush, I think it does make sense to just bundle it up. It's a lot to QA, for sure; but I think we're actually saving QA in the long run.

> Now that you know what I think, it's okay. We can continue fixing all the issues that arise with this PR. I guess at some point everything will be okay.

Yup; we'll incrementally get there.

> In term of the subpixel rendering the new modification looks fine even if maybe not identical to the existing filter.

Sounds good.

> I only noticed the problem with character spacing. It seems to happen only on windows for some reason. I will double check and try to post the image.

OK; please post when you can. I'll look into what you mentioned.

franko commented:
Here the image,

![lite-xl-bad-font-spacing](https://user-images.githubusercontent.com/433545/134913661-273e5563-3068-4cf9-8bbc-1fd51af02573.PNG)

I think for some reason only happen on Windows don't know why. I guess the fact this doesn't seem to happen on Linux could be the reason you didn't notice the problem.

adamharrison commented:
Yeah. I'll go in on windows and see if I can find out what's up.

adamharrison commented:
It's interesting; the metrics are computed totally fine on grayscale rendering. So it's something specifically to do with subpixel rendering. I'll see what I can do about this.

adamharrison commented:
OK! Turned out that you can't just assume things are the same across translated subpixels. Makes sense. Now fixed! I think we're ready to go! Unless you can see any other problems.

franko commented:
Great work @adamharrison. I tested a bit everything seems perfect now.

I am ready to merge but I think you should fix the whitespace plugin before, at least do not show whitespaces by default. Please add a final commit to fix this minor thing.

The other problem I see is that I would prefer to keep this change to be included after the 2.0.3 release as this is a major change. It changes slightly the font rendering and sizing and despite our tests we may introduce some regressions.

I propose to create a branch for the 2.0.3 release from the current master and merge this PR into the master.

What are your thoughts about that ?

adamharrison commented:
Yup, agreed. Sounds good.

adamharrison commented:
OK, added a line to turn off drawing whitespace by default.

adamharrison commented:
Oh, and also added in your proposed changed to synchronise the scales.

franko commented:
The change is now in the master branch, thank you again @adamharrison for this great work.
